File: prakticke/pu02/cpp/FormulaTest.cpp
#include "FormulaTypes.h"
#include "Formula.h"

struct ValCase {
	Valuation valuation;
	bool result;
};

struct Case {
	Formula *formula;
	std::string string;
	int deg;
	Variables vars;
	std::vector<ValCase> vals;
};

class Tester {
	int m_tested = 0;
	int m_passed = 0;
public:
	template<typename T>
	void compare(const T &result, const T &expected, const std::string &msg)
	{
		m_tested++;
		if (result == expected) {
			m_passed++;
		}
		else {
			std::cerr << "    Failed: " << msg << ":" << std::endl;
			std::cerr << "      got " << result << " expected: " << expected << std::endl;
		}
	}

	void compareF(const Formula *result, const Formula *expected, const std::string &msg)
	{
		m_tested++;
		if (result->equals(expected)) {
			m_passed++;
		}
		else {
			std::cerr << "    Failed: " << msg << ":" << std::endl;
			std::cerr << "      got " << result->toString() << " expected: " << expected->toString() << std::endl;
		}
	}

	void testToString(const Case &c) {
		compare(c.formula->toString(), c.string, "toString " + c.string);
	}

	void testDeg(const Case &c) {
		compare(c.formula->deg(), c.deg, "deg " + c.string);
	}

	void testVars(const Case &c) {
		compare(c.formula->vars(), c.vars, "vars " + c.string);
	}

	void testIsSatisfied(const Case &c)
	{
		for (const auto &vc : c.vals) {
			compare(c.formula->isSatisfied(vc.valuation), vc.result,
				"isSatisfied " + c.string + " in " + vc.valuation);
		}
	}

	void test(Formula *f, std::string str, const std::vector<ValCase> &cases)
	{
		std::cerr << "Testing " << str << std::endl;
		compare(f->toString(), str, "toString");
		for (const auto &c : cases) {
			std::cerr << "  Valuation " << c.valuation << std::endl;
			compare(f->isSatisfied(c.valuation), c.result, "isSatisfied");
		}
		delete f;
	}

	bool status()
	{
		std::cerr << "TESTED " << m_tested << std::endl;
		std::cerr << "PASSED " << m_passed << std::endl;
		std::cerr << ( m_tested == m_passed ? "OK" : "ERROR" ) << std::endl;
		return m_tested == m_passed;
	}
};


int main()
{
	Tester t;

	Valuation va1, va2;
	va1["a"] = true;  va2["a"] = false;

	std::vector<Valuation> valuations2;
	{ Valuation v; v["a"] = false; v["b"] = false ; valuations2.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = true  ; valuations2.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = false ; valuations2.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = true  ; valuations2.push_back(v); }

	std::vector<Valuation> valuations3;
	{ Valuation v; v["a"] = false; v["b"] = false, v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = false, v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = true , v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = true , v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = false, v["c"] = true  ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = false, v["c"] = true  ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = true , v["c"] = true  ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = true , v["c"] = true  ; valuations3.push_back(v); }


	std::vector<Case> cases{
		{ new Variable("a"), "a", 0, {"a"},
			{{va1,true}, {va2, false}}
		},
		{ new Negation(new Variable("a")), "-a", 1, {"a"},
			{
				ValCase{va1, false},
				ValCase{va2, true},
			}
		},
		{ new Conjunction( { new Variable("a"), new Variable("b") } ),
			"(a&b)", 1, {"a", "b"},
			{
				ValCase{valuations2[0], false},
				ValCase{valuations2[1], false},
				ValCase{valuations2[2], false},
				ValCase{valuations2[3], true},
			}
		},

		{ new Disjunction( { new Variable("a"), new Variable("b") } ),
			"(a|b)", 1, {"a", "b"},
			{
				ValCase{valuations2[0], false},
				ValCase{valuations2[1], true},
				ValCase{valuations2[2], true},
				ValCase{valuations2[3], true},
			}
		},
		{ new Implication( new Variable("a"), new Variable("b") ),
			"(a->b)", 1, {"a", "b"},
			{
				ValCase{valuations2[0], true},
				ValCase{valuations2[1], true},
				ValCase{valuations2[2], false},
				ValCase{valuations2[3], true},
			}
		},
		{ new Equivalence( new Variable("a"), new Variable("b") ),
			"(a<->b)", 1, {"a", "b"},
			{
				ValCase{valuations2[0], true},
				ValCase{valuations2[1], false},
				ValCase{valuations2[2], false},
				ValCase{valuations2[3], true},
			}
		},
		{ new Disjunction({
				new Negation(new Implication(new Variable("a"),new Variable("b"))),
				new Negation(new Implication(new Variable("b"),new Variable("a")))
			}),
			"(-(a->b)|-(b->a))", 5, {"a", "b"},
			{
				ValCase{valuations2[0], false},
				ValCase{valuations2[1], true},
				ValCase{valuations2[2], true},
				ValCase{valuations2[3], false},
			}
		},
		{ new Conjunction({
				new Implication(new Variable("a"),new Variable("b")),
				new Implication(new Negation(new Variable("a")),new Variable("c"))
			}),
			"((a->b)&(-a->c))", 4, {"a", "b", "c"},
			{
				ValCase{valuations3[0], false},
				ValCase{valuations3[1], false},
				ValCase{valuations3[2], false},
				ValCase{valuations3[3], true},
				ValCase{valuations3[4], true},
				ValCase{valuations3[5], false},
				ValCase{valuations3[6], true},
				ValCase{valuations3[7], true},
			}
		},
		{ new Equivalence(
				new Conjunction({
					new Variable("a"),
					new Negation(new Variable("b"))
				}),
				new Disjunction({
					new Variable("a"),
					new Implication(
						new Variable("b"),
						new Variable("a")
					)
				})
			),
			"((a&-b)<->(a|(b->a)))", 5, {"a", "b"},
			{
				ValCase{valuations2[0], false},
				ValCase{valuations2[1], true},
				ValCase{valuations2[2], true},
				ValCase{valuations2[3], false},
			}
		},
	};

	std::cerr << "Testing toString\n";
	for (const auto &c : cases)
		t.testToString(c);

	std::cerr << "Testing deg\n";
	for (const auto &c : cases)
		t.testDeg(c);

	std::cerr << "Testing vars\n";
	for (const auto &c : cases)
		t.testVars(c);

	std::cerr << "Testing isSatisfied\n";
	for (const auto &c : cases)
		t.testIsSatisfied(c);

	{
		std::cerr << "Testing Negation.originalFormula" << std::endl;
		Formula *a = new Variable("a");
		Formula *na = new Negation(a);
		Negation *nna = new Negation(na);
		t.compare(nna->originalFormula(), na, "Negation.originalFormula");
		delete nna;
	}

	{
		std::cerr << "Testing Implication rightSide / leftSide" << std::endl;
		Formula *a = new Variable("a");
		Formula *b = new Variable("b");
		Formula *na = new Negation(a);
		Implication *nab = new Implication(na, b);
		t.compare(nab->leftSide(), na, "Implication.leftSide");
		t.compare(nab->rightSide(), b, "Implication.rightSide");
		delete nab;
	}

	{
		std::cerr <<  "Testing Equivalence rightSide / leftSide" << std::endl;
		Formula *a = new Variable("a");
		Formula *b = new Variable("b");
		Formula *na = new Negation(a);
		Equivalence *nab = new Equivalence(na, b);
		t.compare(nab->leftSide(), na, "Equivalence.leftSide");
		t.compare(nab->rightSide(), b, "Equivalence.rightSide");
		delete nab;
	}

	std::cerr << "Testing Equals\n";
	{
		Formula *a1 = new Variable("a");
		Formula *a2 = new Variable("a");
		Formula *b = new Variable("b");
		t.compare(a1->equals(a1), true, "equals a == itself");
		t.compare(a1->equals(a2), true, "equals a == a");
		t.compare(a1->equals(b), false, "equals a != b");
		delete a1; delete a2; delete b;
	}
	{
		Formula *a1 = new Negation(new Variable("a"));
		Formula *a2 = new Negation(new Variable("a"));
		Formula *b = new Negation(new Variable("b"));
		t.compare(a1->equals(a1), true, "equals -a == itself");
		t.compare(a1->equals(a2), true, "equals -a == -a");
		t.compare(a1->equals(b), false, "equals -a != -b");
		delete a1; delete a2; delete b;
	}
	{
		Formula *a1 = new Conjunction({new Variable("a"), new Variable("b")});
		Formula *a2 = new Conjunction({new Variable("a"), new Variable("b")});
		Formula *b1 = new Conjunction({new Variable("c"), new Variable("b")});
		Formula *b2 = new Conjunction({new Variable("a"), new Variable("d")});
		t.compare(a1->equals(a1), true, "equals (a&b) == itself");
		t.compare(a1->equals(a2), true, "equals (a&b) == (a&b)");
		t.compare(a1->equals(b1), false, "equals (a&b) != (c&b)");
		t.compare(a1->equals(b2), false, "equals (a&b) != (a&d)");
		t.compare(b1->equals(b2), false, "equals (c&b) != (a&d)");
		delete a1; delete a2; delete b1; delete b2;
	}
	{
		Formula *a1 = new Disjunction({new Variable("a"), new Variable("b")});
		Formula *a2 = new Disjunction({new Variable("a"), new Variable("b")});
		Formula *b1 = new Disjunction({new Variable("c"), new Variable("b")});
		Formula *b2 = new Disjunction({new Variable("a"), new Variable("d")});
		t.compare(a1->equals(a1), true, "equals (a|b) == itself");
		t.compare(a1->equals(a2), true, "equals (a|b) == (a|b)");
		t.compare(a1->equals(b1), false, "equals (a|b) != (c|b)");
		t.compare(a1->equals(b2), false, "equals (a|b) != (a|d)");
		t.compare(b1->equals(b2), false, "equals (c|b) != (a|d)");
		delete a1; delete a2; delete b1; delete b2;
	}
	{
		Formula *a1 = new Implication(new Variable("a"), new Variable("b"));
		Formula *a2 = new Implication(new Variable("a"), new Variable("b"));
		Formula *b1 = new Implication(new Variable("c"), new Variable("b"));
		Formula *b2 = new Implication(new Variable("a"), new Variable("d"));
		t.compare(a1->equals(a1), true, "equals (a->b) == itself");
		t.compare(a1->equals(a2), true, "equals (a->b) == (a->b)");
		t.compare(a1->equals(b1), false, "equals (a->b) != (c->b)");
		t.compare(a1->equals(b2), false, "equals (a->b) != (a->d)");
		t.compare(b1->equals(b2), false, "equals (c->b) != (a->d)");
		delete a1; delete a2; delete b1; delete b2;
	}
	{
		Formula *a1 = new Implication(new Variable("a"), new Variable("b"));
		Formula *a2 = new Implication(new Variable("a"), new Variable("b"));
		Formula *b1 = new Implication(new Variable("c"), new Variable("b"));
		Formula *b2 = new Implication(new Variable("a"), new Variable("d"));
		t.compare(a1->equals(a1), true, "equals (a<->b) == itself");
		t.compare(a1->equals(a2), true, "equals (a<->b) == (a<->b)");
		t.compare(a1->equals(b1), false, "equals (a<->b) != (c<->b)");
		t.compare(a1->equals(b2), false, "equals (a<->b) != (a<->d)");
		t.compare(b1->equals(b2), false, "equals (c<->b) != (a<->d)");
		delete a1; delete a2; delete b1; delete b2;
	}
	{
		Formula *a = new Equivalence(
				new Conjunction({new Variable("a"), new Negation(new Variable("b"))}),
				new Disjunction({new Variable("a"), new Implication(new Variable("b"), new Variable("a"))})
		);
		Formula *b = new Equivalence(
				new Conjunction({ new Variable("a"), new Negation(new Variable("b"))}),
				new Disjunction({ new Variable("a"), new Implication(new Variable("b"), new Variable("a"))})
		);
		Formula *c = new Equivalence(
				new Disjunction({new Variable("a"),new Negation(new Variable("b"))}),
				new Disjunction({new Variable("a"),new Implication(new Variable("b"),new Variable("a"))})
		);
		t.compare(a->equals(a), true, "equals ((a&-b)<->(a|(b->a))) == itself" );
		t.compare(a->equals(b), true, "equals ((a&-b)<->(a|(b->a))) == ((a&-b)<->(a|(b->a)))");
		t.compare(a->equals(c), false, "equals ((a&-b)<->(a|(b->a))) != ((a|-b)<->(a|(b->a)))");
		delete a; delete b; delete c;
	}

	{
		std::cerr << "Testing substitute\n";
		Formula *a = new Equivalence(
				new Conjunction({new Variable("a"), new Negation(new Variable("b"))}),
				new Disjunction({new Variable("a"), new Implication(new Variable("b"), new Variable("a"))})
		);
		Formula *b = new Equivalence(
				new Disjunction({new Variable("a"),new Negation(new Variable("b"))}),
				new Disjunction({new Variable("a"),new Implication(new Variable("b"),new Variable("a"))})
		);
		Formula *w = new Conjunction({new Variable("a"), new Negation(new Variable("b"))});
		Formula *r = new Disjunction({new Variable("a"), new Negation(new Variable("b"))});
		Equivalence *s = dynamic_cast<Equivalence*>(a->substitute(w, r));
		assert(s);
		t.compare(s->equals(b), true, "substitute " + r->toString() + " for " + w->toString() + " in " + a->toString() + " -> ((a|-b)<->(a|(b->a)))");
		t.compare(s == a, false, "substite creates new copy");

		t.compare(s->leftSide()->equals(r), true, "substitute replaces with equal formula");
		t.compare(s->leftSide() == r, false, "substitute replaces with new copy");

		Formula *r2 = r->substitute(new Variable("b"), new Variable("c"));
		Formula *d = new Disjunction({new Variable("a"), new Negation(new Variable("c"))});
		t.compare(r2->equals(d), true, "substitute b for c in (a|-b) -> (a|-c)");
		delete a; delete b; delete w; delete r; delete s; delete r2; delete d;
	}

	return t.status() ? 0 : 1;
}

File: prakticke/pu02/java/FormulaTest.java
import java.lang.System;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.Arrays;

class ValCase {
    Map<String,Boolean> valuation;
    boolean result;

    ValCase(Map<String, Boolean> valuation, boolean result) {
        this.valuation = valuation;
        this.result    = result;
    }
}

class Case {
    Formula formula;
    String string;
    int deg;
    Set<String> vars;
    ValCase[] vals;

    Case(Formula f, String s, int d, Set<String> vars, ValCase[] vals){
        this.formula = f;
        this.string = s;
        this.deg = d;
        this.vars = vars;
        this.vals = vals;
    }
}

class Tester {
    int tested = 0;
    int passed = 0;

    public void compare(Object result, Object expected, String msg) {
        tested += 1;
        if (result.equals(expected)) {
            passed += 1;
        } else {
            System.out.println("    Failed: " + msg + ":");
            System.out.println("      got " + result + " expected " + expected);
        }
    }

    public void compare(Formula result, Formula expected, String msg) {
        tested += 1;
        if (result.equals(expected)) {
            passed += 1;
        } else {
            System.out.println("    Failed: " + msg + ":");
            System.out.println("      got " + result + " expected " + expected);
        }
    }

    public void compareInstance(Object result, Object expected, String msg) {
        tested += 1;
        if (result == expected) {
            passed += 1;
        } else {
            System.out.println("    Failed compareInstance: " + msg + ":");
            System.out.println("      got " + result + " expected " + expected);
        }
    }

    public void testToString(Case c) {
        compare(c.formula.toString(), c.string, "toString " + c.string);
    }

    public void testDeg(Case c) {
        compare(c.formula.deg(), c.deg, "deg " + c.string);
    }

    public void testVars(Case c) {
        compare(c.formula.vars(), c.vars, "vars " + c.string);
    }

    public void testIsSatisfied(Case c) {
        for (ValCase vc : c.vals) {
            compare(c.formula.isSatisfied(vc.valuation), vc.result, "isSatisfied " + c.string + " in " + vc.valuation);
        }
    }

    public void test(Formula formula, String string, ValCase[] cases) {
        System.out.println("Testing " + string);
        compare(formula.toString(), string, "toString");
        for (ValCase c : cases) {
            System.out.println(" valuation " + c.valuation + ":");
            compare(formula.isSatisfied(c.valuation), c.result, "isSatisfied");
        }
    }

    public boolean status() {
        System.out.println("TESTED " + tested + "\nPASSED " + passed);
        if (tested == passed) {
            System.out.println("OK");
            return true;
        } else {
            System.out.println("ERROR");
            return false;
        }
    }
}

public class FormulaTest {

    static Variable Var(String v) { return new Variable(v); }
    static Negation Neg(Formula f) { return new Negation(f); }
    static Conjunction And(Formula[] fs) { return new Conjunction(fs); }
    static Disjunction Or(Formula[] fs) { return new Disjunction(fs); }
    static Implication Impl(Formula l, Formula r) { return new Implication(l, r); }
    static Equivalence Eq(Formula l, Formula r) { return new Equivalence(l, r); }
    static Case CS(Formula f, String s, int d, Set<String> vars, ValCase[] vals) { return new Case(f, s, d, vars, vals); }
    static ValCase Val(Map<String, Boolean> v, boolean r) { return new ValCase(v, r); }

    public static void main(String[] args) {
        Tester t = new Tester();

        Set<String> ss = new HashSet<String>(Arrays.asList("a", "b"));

        Map<String,Boolean> v1 = new HashMap<>();
        v1.put("a", true);

        Map<String,Boolean> v2 = new HashMap<>();
        v2.put("a", false);

        Map<String,Boolean> a1 = new HashMap<>();
        a1.put("a", false);
        a1.put("b", false);

        Map<String, Boolean> a2 = new HashMap<>();
        a2.put("a", false);
        a2.put("b", true);

        Map<String, Boolean> a3 = new HashMap<>();
        a3.put("a", true);
        a3.put("b", false);

        Map<String, Boolean> a4 = new HashMap<>();
        a4.put("a", true);
        a4.put("b", true);

        Map<String,Boolean> b1 = new HashMap<>();
        b1.put("a", false);
        b1.put("b", false);
        b1.put("c", false);

        Map<String,Boolean> b2 = new HashMap<>();
        b2.put("a", true);
        b2.put("b", false);
        b2.put("c", false);

        Map<String,Boolean> b3 = new HashMap<>();
        b3.put("a", false);
        b3.put("b", true);
        b3.put("c", false);

        Map<String,Boolean> b4 = new HashMap<>();
        b4.put("a", true);
        b4.put("b", true);
        b4.put("c", false);

        Map<String,Boolean> b5 = new HashMap<>();
        b5.put("a", false);
        b5.put("b", false);
        b5.put("c", true);

        Map<String,Boolean> b6 = new HashMap<>();
        b6.put("a", true);
        b6.put("b", false);
        b6.put("c", true);

        Map<String,Boolean> b7 = new HashMap<>();
        b7.put("a", false);
        b7.put("b", true);
        b7.put("c", true);

        Map<String,Boolean> b8 = new HashMap<>();
        b8.put("a", true);
        b8.put("b", true);
        b8.put("c", true);

        Case[] cases = new Case[] {
            CS(new Variable("a"), "a", 0,
                new HashSet<String>(Arrays.asList("a")),
                new ValCase[] {
                Val(v1, true), Val(v2, false)
            }),

            new Case(new Negation(new Variable("a")), "-a", 1,
                new HashSet<String>(Arrays.asList("a")),
                new ValCase[] {
                Val(v1, false), Val(v2, true) 
            }),
            new Case(new Conjunction( new Formula[] { new Variable("a"), new Variable("b") } ), "(a&b)", 1,
                new HashSet<String>(Arrays.asList("a", "b")),
                new ValCase[] {
                Val(a1, false), Val(a2, false), Val(a3, false), Val(a4, true)
            }),
            new Case(new Disjunction( new Formula[] { new Variable("a"), new Variable("b") } ), "(a|b)", 1,
                new HashSet<String>(Arrays.asList("a", "b")),
                new ValCase[] {
                Val(a1, false), Val(a2, true), Val(a3, true), Val(a4, true)
            }),
            new Case(new Implication( new Variable("a"), new Variable("b") ), "(a->b)", 1,
                new HashSet<String>(Arrays.asList("a", "b")),
                new ValCase[] {
                Val(a1, true), Val(a2, true), Val(a3, false), Val(a4, true)
            }),
            new Case(new Equivalence( new Variable("a"), new Variable("b") ), "(a<->b)", 1,
                new HashSet<String>(Arrays.asList("a", "b")),
                new ValCase[] {
                new ValCase(a1, true), new ValCase(a2, false), new ValCase(a3, false), new ValCase(a4, true)
            }),
            new Case(new Disjunction(new Formula[] {
                    new Negation(new Implication(new Variable("a"), new Variable("b"))),
                    new Negation(new Implication(new Variable("b"), new Variable("a")))
                }), "(-(a->b)|-(b->a))", 5, 
                new HashSet<String>(Arrays.asList("a", "b")),
                new ValCase[] {
                new ValCase(a1, false), new ValCase(a2, true), new ValCase(a3, true), new ValCase(a4, false)
            }),
            new Case(new Conjunction(new Formula[] {
                    new Implication(new Variable("a"), new Variable("b")),
                    new Implication(new Negation(new Variable("a")), new Variable("c"))
                }), "((a->b)&(-a->c))", 4,
                new HashSet<String>(Arrays.asList("a", "b", "c")),
                new ValCase[] {
                    new ValCase(b1, false), new ValCase(b2, false), new ValCase(b3, false),
                    new ValCase(b4, true), new ValCase(b5, true), new ValCase(b6, false),
                    new ValCase(b7, true), new ValCase(b8, true)
            }),
            new Case(new Equivalence(
                    new Conjunction(new Formula[] { new Variable("a"), new Negation(new Variable("b")) }),
                    new Disjunction(new Formula[] { new Variable("a"), new Implication( new Variable("b"), new Variable("a")) })
                ), "((a&-b)<->(a|(b->a)))", 5,
                new HashSet<String>(Arrays.asList("a", "b")),
                new ValCase[] {
                    new ValCase(a1, false), new ValCase(a2, true), new ValCase(a3, true), new ValCase(a4, false)
            }),
        };

        System.out.println("Testing toString");
        for (Case c : cases) {
            t.testToString(c);
        }

        System.out.println("Testing deg");
        for (Case c : cases) {
            t.testDeg(c);
        }

        System.out.println("Testing vars");
        for (Case c : cases) {
            t.testVars(c);
        }

        System.out.println("Testing isSatisfied");
        for (Case c : cases) {
            t.testIsSatisfied(c);
        }

        {
            System.out.println("Testing Variable.name");
            Variable a = new Variable("a");
            Variable b = new Variable("beta");
            t.compare(a.name(), "a", "Variable.name");
            t.compare(b.name(), "beta", "Variable.name");
        }

        {
            System.out.println("Testing Negation.originalFormula");
            Formula a = new Variable("a");
            Formula na = new Negation(a);
            Negation nna = new Negation(na);
            t.compareInstance(nna.originalFormula(), na, "Negation.originalFormula");
        }

        {
            System.out.println("Testing Implication rightSide / leftSide");
            Formula a = new Variable("a");
            Formula b = new Variable("b");
            Formula na = new Negation(a);
            Implication nab = new Implication(na, b);
            t.compareInstance(nab.leftSide(), na, "Implication.leftSide");
            t.compareInstance(nab.rightSide(), b, "Implication.rightSide");
        }

        {
            System.out.println("Testing Equivalence rightSide / leftSide");
            Formula a = new Variable("a");
            Formula b = new Variable("b");
            Formula na = new Negation(a);
            Equivalence nab = new Equivalence(na, b);
            t.compareInstance(nab.leftSide(), na, "Equivalence.leftSide");
            t.compareInstance(nab.rightSide(), b, "Equivalence.rightSide");
        }

        {
            System.out.println("Testing Equals");
            t.compare(Var("a").equals(Var("a")), true, "equals");
            t.compare(Var("a").equals(Var("b")), false, "equals");
            Formula a = Eq(
                    And(new Formula[] {Var("a"), Neg(Var("b"))}),
                    Or(new Formula[] {Var("a"), Impl(Var("b"), Var("a"))})
            );
            Formula b = Eq(
                    And(new Formula[] { Var("a"), Neg(Var("b"))}),
                    Or(new Formula[] {Var("a"),Impl(Var("b"),Var("a"))})
            );
            Formula c = Eq(
                    Or(new Formula[] {Var("a"),Neg(Var("b"))}),
                    Or(new Formula[] {Var("a"),Impl(Var("b"),Var("a"))})
            );
            t.compare(a.equals(a), true, "equals");
            t.compare(a.equals(b), true, "equals");
            t.compare(a.equals(c), false, "equals");
        }
        {
            System.out.println("Testing substitute");
            Formula a = Eq(
                    And(new Formula[] {Var("a"), Neg(Var("b"))}),
                    Or(new Formula[] {Var("a"), Impl(Var("b"), Var("a"))})
            );
            Formula b = Eq(
                    And(new Formula[] { Var("a"), Neg(Var("b"))}),
                    Or(new Formula[] {Var("a"),Impl(Var("b"),Var("a"))})
            );
            Formula c = Eq(
                    Or(new Formula[] {Var("a"),Neg(Var("b"))}),
                    Or(new Formula[] {Var("a"),Impl(Var("b"),Var("a"))})
            );
            Formula w = And(new Formula[] {Var("a"), Neg(Var("b"))});
            Formula r = Or(new Formula[] {Var("a"), Neg(Var("b"))});
            Equivalence s = (Equivalence)a.substitute(w, r);
            t.compare(s, c, "substitute " + r + " for " + w + " in " + a + " -> ((a|-b)<->(a|(b->a)))");
            t.compare(s == a, false, "substite creates new copy");

            t.compare(s.leftSide(), r, "substitute replaces with equal formula");
            t.compare(s.leftSide() == r, false, "substitute replaces with new copy");

            Formula r2 = r.substitute(Var("b"), Var("c"));
            t.compare(r2, Or(new Formula[] {Var("a"), Neg(Var("c"))}),
                    "substitute b for c in (a|-b) -> (a|-c)");
        }

        System.exit(t.status() ? 0 : 1);
    }
}
